docs/design/ha_master.md

```diff
@@ -0,0 +1,188 @@
+in front of apiservers we will have an L4 load balancer to spread the load among all the replicas.
+* Controller manager, scheduler & cluster autoscaler will use lease mechanism and
+only a single instance will be an active master. All other will be waiting in a standby mode.
+* All add-on managers will work independently and each of them will try to keep add-ons in sync
+
+# Detailed design
+
+## VM naming
+VMs with master replicas will have the following names:
+* first VM - `${MASTER_NAME}`
+* next VMs - `${MASTER_NAME}-<hash>`, where `<hash>` is a 3-characters long alphanumeric string.
+
+It will be ok to delete VM without the suffix.
+
+If user will run deployment scripts with a different `MASTER_NAME` variable than earlier,
+it will fail with an error that there are no masters to replicate.
+
+##Components
+
+###etcd
+```Note: this paragraphs assumes we are using etcd v2.2; it will have to be revisited during
+upgrade to v3, e.g. different command line options.```
+
+All etcd instances will be clustered together and one of them will be an elected master.
+In order to commit any change quorum of the cluster will have to confirm it. Etcd will be
+configured in such a way that all writes and reads will go through the master (requests
+will be forwarded by the local etcd server such that it’s invisible for the user). It will
+affect latency for all operations but it should not increase by much more than the network
+latency between master replicas (latency between zones with a region is < 10ms).
+
+Currently etcd exposes port only using localhost interface. In order to allow clustering
+and inter-VM communication we will also have to use public interface. To secure the
+communication we will configure firewall rules to only allow traffic on etcd port from
+other master machines.
+
+Another option to secure etcd cluster communication would be to use SSL (as described
+(here)[https://coreos.com/etcd/docs/latest/security.html]), but it would increase CPU usage
+on master machines, which is already a bottleneck in large clusters. As there is no clear
+benefits we decided to use a simpler and faster solution described above.
+
+When generating command line for etcd we will always assume it’s part of a cluster
+(initially of size 1) and list all VMs for kubernetes masters (see VM naming section).
+Based on that, we will set the following flags:
+* `-initial-cluster` - list of all hostnames for master machines
+* `-initial-cluster-state` (keep in mind that we are adding masters one by one):
+  * `new` if we are adding the first replica, i.e. the list of existing masters is empty
+  * `existing` if there are more than one replica, i.e. the list of existing masters is non-empty.
+
+This will allow us to have exactly the same logic for HA and non-HA master. List of VMs
+with master replicas will be generated in `kube-up.sh` script and passed to as a env variable
+`INITIAL_ETCD_CLUSTER`.
```

Comment:
It bootstraps a cluster with a fixed number of members, and deals with how you operate that cluster even as cluster members are failing, in the scenario where you can remap storage (e.g. GCE PD volumes).  The approach allows for "no-ops" when maintaining an etcd cluster.

It does not yet address dynamically reconfiguring the cluster, that is true.   But I suggest that we need to figure out how to take this recommended approach to running a cluster and extend it to include membership changes.  I believe the static configuration is only used for first cluster boot, so I suspect we will bring up the first cluster (maybe with N=1) following the recommended approach, and then add additional members consistently - i.e. with DNS names that we remap.

The problem that we risk falling into otherwise is that we're baking in a GCE assumption.  GCE gives you pet-like reliable instances with fixed internal IPs, that will be rebooted and migrated.  AWS "officially" supports something similar, but it is really not a good idea in practice - it is much safer on AWS to use auto-scaling groups.  And I think other clouds do _not_ offer you control over IP allocation, for example.